components/script/dom/cssstyledeclaration.rs

```diff
@@ -235,21 +238,88 @@ impl CSSStyleDeclaration {
     }
 
     fn get_computed_style(&self, property: PropertyId) -> DOMString {
-        match self.owner {
+        let node = match self.owner {
             CSSStyleOwner::CSSRule(..) => {
                 panic!("get_computed_style called on CSSStyleDeclaration with a CSSRule owner")
             },
-            CSSStyleOwner::Element(ref el) => {
-                let node = el.upcast::<Node>();
-                if !node.is_connected() {
-                    // TODO: Node should be matched against the style rules of this window.
-                    // Firefox is currently the only browser to implement this.
-                    return DOMString::new();
-                }
-                let addr = node.to_trusted_node_address();
-                window_from_node(node).resolved_style_query(addr, self.pseudo.clone(), property)
+            CSSStyleOwner::Element(ref element) => element.upcast::<Node>(),
+        };
+
+        if !node.is_connected() {
+            // TODO: Node should be matched against the style rules of this window.
+            // Firefox is currently the only browser to implement this.
+            return DOMString::new();
+        }
+
+        if let Some(result) = self.try_to_get_resolved_style(node, &property) {
+            return result.into();
+        }
+
+        let addr = node.to_trusted_node_address();
+        window_from_node(node).resolved_style_query(addr, self.pseudo.clone(), property)
+    }
+
+    // Try to get the resolved style for a node without having to ask the layout
+    // thread. This is possible when we can simply return the computed value or
+    // we can calculate the used value without looking at layout data structures.
+    // Some properties resovled values compute are their computed values.
+    // See: https://drafts.csswg.org/cssom/#dom-window-getcomputedstyle
+    fn try_to_get_resolved_style(&self, node: &Node, property: &PropertyId) -> Option<String> {
```

Comment:
This doesn't look good if there are pending style changes. What prevents you from using an outdated style in that case?

Reply:
Answering myself: `.style()` / `.style_for_pseudo` do update style.

Reply:
Why is this optimization valuable then, though? Is the idea that it avoids going through the layout thread if there are no pending changes? If so it should be documented.